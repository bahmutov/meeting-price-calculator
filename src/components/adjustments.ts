import xs, { Stream } from 'xstream';
import { VNode, div, button } from '@cycle/dom';

import { Sources, Sinks } from '../interfaces';

export default function Adjustments(sources : Sources) : Sinks {
    const vdom$ : Stream<VNode> = xs.of(
        div('.Adjustments', [
            button('.dec', 'Decrement'),
            button('.inc', 'Increment'),
        ])
    );

    const action$ : xs<number> = xs.merge(
        sources.DOM.select('.dec').events('click').mapTo(-1),
        sources.DOM.select('.inc').events('click').mapTo(+1)
    );

    const adjustments$ : Stream<number> = action$.fold((x, y) => x + y, 4);

    const sinks : Sinks = {
        DOM : vdom$,
        ADJUSTMENTS: adjustments$
    };

    return sinks;
}
