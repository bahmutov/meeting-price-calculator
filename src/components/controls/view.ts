import xs from 'xstream';
import * as moment from 'moment';
import { VNode, div, select, option, span, label } from '@cycle/dom';

import { State } from './index';
import { styles } from './styles';
import { formatPrice } from '../../utils/priceUtils';

function renderOption(currency: string, selectedCurrency: string) {
  return currency === selectedCurrency
    ? option(
        {
          attrs: {
            selected: 'selected'
          }
        },
        [currency]
      )
    : option([currency]);
}

function renderCurrencySelect(selectedCurrency: string) {
  return select(`.${styles.currencySelect}`, [
    renderOption('â‚¬', selectedCurrency),
    renderOption('$', selectedCurrency)
  ]);
}

export default function view(
  state$: xs<State>,
  personAmountSliderVDom$: xs<VNode>,
  avgPriceSliderVDom$: xs<VNode>
): xs<VNode> {
  return xs
    .combine(state$, personAmountSliderVDom$, avgPriceSliderVDom$)
    .map(
      ([
        { currency, personAmount, avgPrice },
        personAmountVDom,
        avgPriceVDom
      ]) =>
        div(`.${styles.controls}`, [
          personAmountVDom,
          div(`.${styles.priceResult}`, [
            label('Total price per hour'),
            div(
              `.${styles.totalPriceValue}`,
              formatPrice(personAmount * avgPrice, currency)
            ),
            div(`.${styles.currency}`, [
              span('Currency'),
              renderCurrencySelect(currency)
            ])
          ]),
          avgPriceVDom
        ])
    );
}
