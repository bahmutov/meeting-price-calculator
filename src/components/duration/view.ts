import xs from 'xstream';
import * as moment from 'moment';

import { VNode, div, span, button } from '@cycle/dom';
import { State } from './index';
import { styles } from './styles';

export default function view(state$: xs<State>): xs<VNode> {
  return state$.map(({ startTime, duration }) =>
    div([
      div([
        span(`Start time: ${moment(startTime).format('HH:mm:ss')}`),
        button(`.${styles.resetButton}`, 'Reset')
      ]),
      div(`Duration: ${moment.duration(duration, 'seconds').humanize()}`)
    ])
  );
}
