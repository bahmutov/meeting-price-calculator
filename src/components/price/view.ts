import xs from 'xstream';
import { VNode, div, select, option, span, label } from '@cycle/dom';
import { State } from '../../interfaces';
import calculatePrice, { formatPrice } from '../../utils/priceUtils';

function renderPrice(
  personAmountValue: number,
  avgPriceValue: number,
  tick: number,
  currency: string
): VNode {
  const price: number = calculatePrice(personAmountValue, avgPriceValue, tick);
  return div('.Price-actual', [
    div('.Price-label', 'This meeting has cost'),
    div('.Price-value', formatPrice(price, currency))
  ]);
}

export default function view(
  state$: xs<State>,
  personAmountSliderVDom$: xs<VNode>,
  avgPriceSliderVDom$: xs<VNode>
): xs<VNode> {
  return xs
    .combine(state$, personAmountSliderVDom$, avgPriceSliderVDom$)
    .map(
      (
        [
          { tick, personAmount, avgPrice, currency },
          personAmountVDom,
          avgPriceVDom
        ]
      ) =>
        div('.Price', [
          renderPrice(personAmount.value, avgPrice.value, tick, currency),
          div('.PriceInputs', [
            personAmountVDom,
            div('.price-result', [
              label('.total-price-label', 'Total price per hour'),
              div(
                '.total-price-value',
                formatPrice(personAmount.value * avgPrice.value, currency)
              ),
              div('.currency', [
                span('.currency-label.label', 'Currency'),
                select('.currency-select', [option('â‚¬'), option('$')])
              ])
            ]),
            avgPriceVDom
          ])
        ])
    );
}
