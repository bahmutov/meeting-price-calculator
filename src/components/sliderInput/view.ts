import xs from 'xstream';
import { VNode, div, input, span, label } from '@cycle/dom';
import { State } from './index';
import { styles } from './styles';

export default function view(state$: xs<State>): xs<VNode> {
  return state$.map(({ description, unit, min, max, step, value }) =>
    div(`.${styles.sliderInput}`, [
      label(description),
      input(`.SliderInput-input .${styles.numberInput}`, {
        attrs: {
          type: 'number',
          min,
          max,
          step
        },
        props: { value }
      }),
      span(`.${styles.sliderInputUnit}`, unit),
      input('.SliderInput-input', {
        attrs: {
          type: 'range',
          min,
          max,
          step
        },
        props: { value }
      })
    ])
  );
}
