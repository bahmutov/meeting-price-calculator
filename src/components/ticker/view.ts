import xs from 'xstream';
import { VNode, div } from '@cycle/dom';
import { State } from './index';
import { styles } from './styles';
import { formatPrice } from '../../utils/priceUtils';

export default function view(state$: xs<State>): xs<VNode> {
  return state$.map(({ currency, totalPrice }) =>
    div(`.${styles.actualPrice}`, [
      div(`.${styles.priceLabel}`, 'This meeting has cost'),
      div(`.${styles.priceValue}`, formatPrice(totalPrice, currency))
    ])
  );
}
