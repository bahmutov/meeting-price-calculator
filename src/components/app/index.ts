import xs from 'xstream';
import { VNode, div } from '@cycle/dom';
import isolate from '@cycle/isolate';
import * as moment from 'moment';

import { Sources, Sinks } from '../../interfaces';
import model from './model';
import Header from '../header';
import Ticker, { lens as tickerLens } from '../ticker';
import Controls, { lens as controlsLens } from '../controls';
import Duration, { lens as durationLens } from '../duration';
import Footer from '../footer';
import { State as SliderInputState } from '../sliderInput';

export interface State {
  startTime: moment.Moment;
  duration: number;
  currency: string;
  personAmount: SliderInputState;
  avgPrice: SliderInputState;
}

export type Reducer = (prev?: State) => State | undefined;

export default function App(sources: Sources): Sinks {
  const headerVDom$: xs<VNode> = Header();
  const tickerSinks = isolate(Ticker, { onion: tickerLens })(sources);
  const controlsSinks = isolate(Controls, { onion: controlsLens })(sources);
  const durationSinks = isolate(Duration, { onion: durationLens })(sources);
  const footerVDom$: xs<VNode> = Footer();

  const vdom$ = xs
    .combine(
      headerVDom$,
      tickerSinks.DOM,
      controlsSinks.DOM,
      durationSinks.DOM,
      footerVDom$
    )
    .map(([header, ticker, controls, duration, footer]) =>
      div('.App-flex-container', [
        header,
        div('.wrapper-div', [ticker, controls, duration]),
        footer
      ])
    );

  const parentReducer$ = model();

  const reducer$ = xs.merge(
    parentReducer$,
    tickerSinks.onion,
    controlsSinks.onion,
    durationSinks.onion
  );

  const sinks: any = {
    DOM: vdom$,
    onion: reducer$
  };

  return sinks;
}
