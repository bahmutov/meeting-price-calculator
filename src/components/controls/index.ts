import xs from 'xstream';
import isolate from '@cycle/isolate';
import * as moment from 'moment';

import { Sources, Sinks } from '../../interfaces';
import { State as AppState } from '../app';
import model from './model';
import view from './view';
import intent, { Actions } from './intent';
import SliderInput, { personAmountLens, avgPriceLens } from '../sliderInput';

export interface State {
  currency: string;
  personAmount: number;
  avgPrice: number;
}

export type Reducer = (prev?: State) => State | undefined;

export const lens = {
  get: (state: AppState): State => ({
    currency: state.currency,
    personAmount: state.personAmount,
    avgPrice: state.avgPrice
  }),

  set: (state: AppState, childState: State) => ({
    ...state,
    currency: childState.currency,
    personAmount: childState.personAmount,
    avgPrice: childState.avgPrice
  })
};

export default function Controls(sources: Sources): Sinks {
  const actions: Actions = intent(sources.DOM);

  const personAmountSlider: Sinks = isolate(SliderInput, {
    onion: personAmountLens
  })(sources);

  const avgPriceSlider: Sinks = isolate(SliderInput, {
    onion: avgPriceLens
  })(sources);

  const parentReducer$: xs<Reducer> = model(actions);
  const personAmountReducer$: xs<Reducer> = personAmountSlider.onion;
  const avgPriceReducer$: xs<Reducer> = avgPriceSlider.onion;

  const reducer$: xs<Reducer> = xs.merge(
    parentReducer$,
    personAmountReducer$,
    avgPriceReducer$
  );

  const state$: xs<State> = sources.onion.state$;
  const vdom$ = view(state$, personAmountSlider.DOM, avgPriceSlider.DOM);

  const sinks = {
    DOM: vdom$,
    onion: reducer$
  };

  return sinks;
}
