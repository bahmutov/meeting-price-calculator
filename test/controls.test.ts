import xs from 'xstream';
import { forall, integer, assert, Options } from 'jsverify';
import { VNode, div, p } from '@cycle/dom';
import { withTime } from 'cyclejs-test-helpers';
const htmlLooksLike = require('html-looks-like');
const toHtml = require('snabbdom-to-html');
import * as moment from 'moment';

import view from '../src/components/controls/view';
import { State } from '../src/components/controls';
import calculatePrice, { formatPrice } from '../src/utils/priceUtils';

const testOptions: Options = {
  tests: 25,
  quiet: false,
  size: 60 * 60 * 24
};

describe('Controls Component', () => {
  let originalTimeout;

  beforeEach(() => {
    originalTimeout = jasmine.DEFAULT_TIMEOUT_INTERVAL;
    jasmine.DEFAULT_TIMEOUT_INTERVAL = 20000;
  });

  afterEach(() => {
    jasmine.DEFAULT_TIMEOUT_INTERVAL = originalTimeout;
  });

  const personAmountSliderDOM = xs.of(
    div('.SliderInput', [p('personAmountSliderDOM')])
  );
  const avgPriceSliderDOM = xs.of(
    div('.SliderInput', [p('avgPriceSliderDOM')])
  );

  it('should match a snapshot correctly', () => {
    const startTime = moment('2017-06-29T21:19:35.849Z');
    const state$: xs<State> = xs.of({
      currency: '€',
      personAmount: 8,
      avgPrice: 105
    });
    const vdom$: xs<VNode> = view(
      state$,
      personAmountSliderDOM,
      avgPriceSliderDOM
    );
    const html$: xs<{}> = vdom$.map(toHtml);
    expect(html$).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const startTime = moment();
    const expectedHTML = (
      currency: string,
      personAmount: number,
      avgPrice: number
    ) => {
      const totalPricePerHour = formatPrice(personAmount * avgPrice, currency);
      return `
        <div>
          <div class="SliderInput">
            <p>personAmountSliderDOM</p>
          </div>
          <div>
            <label>Total price per hour</label>
            <div>${totalPricePerHour}</div>
            <div>
              <span>Currency</span>
              <select>
                <option>€</option>
                <option>$</option>
              </select>
            </div>
          </div>
          <div class="SliderInput">
            <p>avgPriceSliderDOM</p>
          </div>
        </div>
      `;
    };

    const property = forall(integer, integer, (pa: number, avg: number) =>
      withTime(Time => {
        const state$: xs<State> = xs.of({
          currency: '€',
          personAmount: pa,
          avgPrice: avg
        });

        const vdom$ = view(state$, personAmountSliderDOM, avgPriceSliderDOM);
        const html$ = vdom$.map(toHtml);

        const expected$ = state$.map(({ currency, personAmount, avgPrice }) =>
          expectedHTML(currency, personAmount, avgPrice)
        );

        Time.assertEqual(html$, expected$, htmlLooksLike);
      })
    );

    return assert(property, testOptions);
  });
});
