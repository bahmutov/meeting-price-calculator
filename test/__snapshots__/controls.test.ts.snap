// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls Component should match a snapshot correctly 1`] = `
Stream {
  "_d": false,
  "_dl": Object {},
  "_err": Object {},
  "_ils": Array [],
  "_prod": MapOp {
    "f": [Function],
    "ins": Stream {
      "_d": false,
      "_dl": Object {},
      "_err": Object {},
      "_ils": Array [],
      "_prod": MapOp {
        "f": [Function],
        "ins": Stream {
          "_d": false,
          "_dl": Object {},
          "_err": Object {},
          "_ils": Array [],
          "_prod": Combine {
            "Nc": 0,
            "Nn": 0,
            "ils": Array [],
            "insArr": Array [
              Stream {
                "_d": false,
                "_dl": Object {},
                "_err": Object {},
                "_ils": Array [],
                "_prod": FromArray {
                  "a": Array [
                    Object {
                      "avgPrice": 105,
                      "currency": "â‚¬",
                      "personAmount": 8,
                    },
                  ],
                  "type": "fromArray",
                },
                "_stopID": Object {},
                "_target": Object {},
              },
              Stream {
                "_d": false,
                "_dl": Object {},
                "_err": Object {},
                "_ils": Array [],
                "_prod": FromArray {
                  "a": Array [
                    <div.SliderInput>
                      personAmountSliderDOM
                    </div.SliderInput>,
                  ],
                  "type": "fromArray",
                },
                "_stopID": Object {},
                "_target": Object {},
              },
              Stream {
                "_d": false,
                "_dl": Object {},
                "_err": Object {},
                "_ils": Array [],
                "_prod": FromArray {
                  "a": Array [
                    <div.SliderInput>
                      avgPriceSliderDOM
                    </div.SliderInput>,
                  ],
                  "type": "fromArray",
                },
                "_stopID": Object {},
                "_target": Object {},
              },
            ],
            "out": Object {},
            "type": "combine",
            "vals": Array [],
          },
          "_stopID": Object {},
          "_target": Object {},
        },
        "out": Object {},
        "type": "map",
      },
      "_stopID": Object {},
      "_target": Object {},
    },
    "out": Object {},
    "type": "map",
  },
  "_stopID": Object {},
  "_target": Object {},
}
`;
